### PR TITLE
Project-happy-thougts-api-SandraWallén

### PR DESCRIPTION
**Pull request by sandra-wallen**

### CODE REVIEW

File: server.js [hunk 1]
```diff
@@ -1,26 +1,149 @@
...
+    default: 0
+  },
+  createdAt: {
+    type: Date,
+    default: Date.now
+  },
+  user: {
+    type: String
+  },
+  hashtag: {
```
Comment (sirimatsdotter) [inline server.js:31]: Love that you included this, like the real twitter #loveit :)

File: server.js [hunk 2]
```diff
@@ -1,26 +1,149 @@
...
+    res.json(thoughtsByUser)
+  } catch (error) {
+    res.status(400).json({ error: 'Something went wrong', details: error })
+  }
+})
+
+app.post('/thoughts', async (req, res) => {
+  let tag
+
+  if (req.body.hashtag.charAt(0) === '#') {
```
Comment (sirimatsdotter) [inline server.js:75]: Cool!

**Review by sirimatsdotter — APPROVED** [review]
GREAT job Sandra!

Very ambitious project and I love that you added tags as well and that also implementing users. If you want to take this further, perhaps you could also create a log in for this and connect the users to the data base as well where the users is a separate collection? :)

File: README.md [hunk 3]
```diff
@@ -1,13 +1,41 @@
...
+Posts a new thought
+- ```POST /thoughts/60a960f77c2917002828615c/likes```
+Posts a like to an existing thought
+
+- ```DELETE /thoughts/60a960f77c2917002828615c```
+Deletes a thought
+- ```PATCH /thoughts/60a960f77c2917002828615c```
+Updates a thought
+- ```PUT /thoughts/60a960f77c2917002828615c```
+Updates a thought
```
Comment (KarolCislowski) [inline README.md:36]: Nice documentation 👍🏻

File: server.js [hunk 4]
```diff
@@ -1,26 +1,149 @@
...
-//   PORT=9000 npm start
-const port = process.env.PORT || 8080
-const app = express()
+const thoughtSchema = new mongoose.Schema({
+  message: {
+    type: String,
+    required: true,
+    minlength: 5,
+    maxlength: 140,
+    trim: true
```
Comment (KarolCislowski) [inline server.js:18]: Nicely configured field 👍🏻

File: server.js [hunk 5]
```diff
@@ -1,26 +1,149 @@
...
+app.use(express.json())
 
-// Start defining your routes here
 app.get('/', (req, res) => {
-  res.send('Hello world')
+  res.send('Welcome to Sandras happy thoughts API!')
+})
+
+app.get('/thoughts', async (req, res) => {
+  const allThoughts = await Thought.find().sort({ createdAt: -1 }).limit(20)
```
Comment (KarolCislowski) [inline server.js:46]: 👍🏻

File: server.js [hunk 6]
```diff
@@ -1,26 +1,149 @@
...
+app.get('/thoughts/user/:name', async (req, res) => {
+  const { name } = req.params
+
+  try {
+    const thoughtsByUser = await Thought.find({ user: name }).sort({ createdAt: -1 })
+    res.json(thoughtsByUser)
+  } catch (error) {
+    res.status(400).json({ error: 'Something went wrong', details: error })
+  }
+})
```
Comment (KarolCislowski) [inline server.js:70]: nice 👍🏻

File: server.js [hunk 7]
```diff
@@ -1,26 +1,149 @@
...
+
+app.post('/thoughts', async (req, res) => {
+  let tag
+
+  if (req.body.hashtag.charAt(0) === '#') {
+    tag = req.body.hashtag.slice(1) 
+  } else {
+    tag = req.body.hashtag
+  }
```
Comment (KarolCislowski) [inline server.js:79]: Nice 👍🏻

File: server.js [hunk 8]
```diff
@@ -1,26 +1,149 @@
...
+  } catch (error) {
+    res.status(400).json(error)
+  } 
+})
+
+app.post('/thoughts/:id/likes', async (req, res) => {
+  const { id } = req.params
+
+  try {
+    const updatedThought = await Thought.findByIdAndUpdate(id, { $inc: { hearts: 1 } }, { new: true })
```
Comment (KarolCislowski) [inline server.js:93]: 💯

File: server.js [hunk 9]
```diff
@@ -1,26 +1,149 @@
...
+  try {
+    const updatedThought = await Thought.findOneAndReplace({ _id: id }, req.body, { new: true })
+    if (updatedThought) {
+      res.json(updatedThought)
+    } else {
+      res.status(404).json({ message: 'Not found' })
+    }
+  } catch (error) {
+    res.status(400).json({ message: 'Invalid request', error })
+  }
```
Comment (KarolCislowski) [inline server.js:146]: Remember to restrict access to these endpoints because not all users should have access to these functionalities 🤓

**Review by KarolCislowski — CHANGES_REQUESTED** [review]
Great API but you need to provide a link to the frontend connected to it 😸

**Comment by sandra-wallen** [comment]
> Great API but you need to provide a link to the frontend connected to it 😸

File: README.md (hunk 3 above)
Comment (sandra-wallen) [reply README.md:36]: @KarolCislowski Hi!

Please see link to github repo for frontend - https://github.com/sandra-wallen/project-happy-thoughts
And deployed frontend - https://sandras-happythoughts.netlify.app/

**Review by KarolCislowski — COMMENTED** [review]
Nice APP

**Pull request closed without merging**